Updates the Basic01 test case specification (#1257)
zonemaster/zonemaster: docs/public/specifications/tests/Basic-TP/basic01.md

## Summary

Message Tag                | Level | Arguments                          | Message ID for message tag
:--------------------------|:------|:-----------------------------------|:--------------------------

> B01_CHILD_IS_ALIAS         |NOTICE |domain_child, domain_target, ns_list| "{domain_child}" is not a zone. It is an alias for "{domain_target}". Run a test for "{domain_target}" instead. Returned from name servers "{ns_list}.

Reviewer: ```suggestion
B01_CHILD_IS_ALIAS         |NOTICE |domain_child, domain_target, ns_list| "{domain_child}" is not a zone. It is an alias for "{domain_target}". Run a test for "{domain_target}" instead. Returned from name servers "{ns_list}".
```

B01_INCONSISTENT_DELEGATION|ERROR  |domain_child, domain_parent, ns_ip_list| The name servers for parent zone "{domain_parent}" give inconsistent delegation of "{domain_child}". Returned from name serververs "{ns_ip_list}".
B01_NO_CHILD               |ERROR  | domain_child, domain_super            | "{domain_child}" does not exist as a DNS zone. Try to test "{domain_super}" instead.
B01_PARENT_FOUND           |INFO   | domain, ns_ip_list                    | The parent zone is "{domain}" as returned from name servers "{ns_ip_list}".
B01_PARENT_NOT_FOUND       |WARNING|                                       | The parent zone cannot be found.
B01_PARENT_UNDETERMINED    |WARNING| ns_ip_list                            | The parent zone cannot be determined on name servers "{ns_ip_list}".

> B01_SERVER_ZONE_ERROR      |DEBUG  | query_name, rrtype, ns_ip             | Unexpected response on "dig {query_name} {rrtype} @{ns_ip}"

Reviewer: I don’t think we should refer to `dig` in our messages. And “unexpected” doesn’t clarify the problem with the response.

Maybe: “Name server {ns_ip} gave an erroneous response to a query of type {query_type} for {query_name}”? Or some variant? This wording isn’t perfect, though, because this message is also emitted when the server fails to respond or responds with NOERROR but with the AA bit unset.

The argument names in the Arguments column lists the arguments used in the
message. The argument names are defined in the [argument list].

The name server names are assumed to be available at the time when the msgid
is created, if the argument name is "ns" or "ns_list" even when in the
"[Test procedure]" below it is only referred to the IP address of the name

> servers.

Reviewer: I’m skeptical about including wording like this. I don’t think it’s a good idea to overload “IP address” to really mean “name and IP address pair”.

Author: That is not what is says. It does not say that "IP adress" means namn and IP. It says that the name is available when creating the messagesl. 

 The text is copied from the end of section https://github.com/zonemaster/zonemaster/blob/master/docs/public/specifications/tests/Nameserver-TP/nameserver15.md#summary

What is you suggestion? Reverting to ns_ip and ns_ip_list, respectively?

   2. Exit the test case.

2. Create [DNS queries][DNS Query]:
   1. Query type DNAME and query name *Child Zone* ("DNAME Child Query").

> 3. Create the following empty sets:

Reviewer: For error reporting’s sake, I’d suggest having sets of (name server name, name server IP, zone) tuples instead of merely (name server IP, zone). Having the name server’s name in addition to its IP address in the error messages makes remediating problems much easier.

Author: Agree. We should move from ns_ip to ns. Updated.

   *Remaining Servers* set.

> In the loop below, the steps tries to capture the name of the parent zone of
> **Child Zone** and the IP addresses of the name servers for that parent zone.
> This is done using a modified version of the "QNAME minimization" technique
> [RFC 9156]. SOA as the query type used for traversing the tree.

5. While the *Remaining Servers* is non-empty pick next name server IP address
   and zone name from the set ("Server Address" and "Zone Name") and do:

   1.  Extract and remove *Server Address* including its *Zone Name* from
       *Remaining Servers*.
   2.  Insert *Server Address* and *Zone Name* into *Handled Servers*.
   3.  Create [DNS queries][DNS Query]:
       1. Query type SOA and query name *Zone Name* ("Zone Name SOA Query").
       2. Query type NS and query name *Zone Name* ("Zone Name NS Query").
   4.  [Send] *Zone Name SOA Query* to *Server Address*.
   5.  Output [B01_SERVER_ZONE_ERROR] with query name *Zone Name*, [query type]
       SOA and name server IP *Server Address* and go to next server in
       *Remaining Servers* if one or more of the following matches:
          * No DNS response.

>           * [RCODE Name] besides NoError in response.

Reviewer: ```suggestion
          * [RCODE Name] different from NoError in response.
```

          * AA bit not set in response.

>           * No SOA record in answer section

Reviewer: What if:

- there are two or more SOA records in the answer section?
- there are two SOA records, one whose owner name is (case-insensitively) equal to the QNAME and one whose owner name is different?

          * Owner name of SOA record is not *Zone Name*.
   6.  [Send] *Zone Name NS Query* to *Server Address*.
   7.  Output [B01_SERVER_ZONE_ERROR] with query name *Zone Name*, [query type]

>        SOA and name server IP *Server Address* and go to next server in

Reviewer: ```suggestion
       NS and name server IP *Server Address* and go to next server in
```

       *Remaining Servers* if one or more of the following matches:
          * No DNS response.

>           * [RCODE Name] besides NoError in response.

Reviewer: ```suggestion
          * [RCODE Name] different from NoError in response.
```

          * AA bit not set in response.
          * No NS records in answer section

>           * Owner name of NS records is not *Zone Name*.

Reviewer: What if the answer section mixes NS records with the owner name matching the QNAME and others that don’t? Shouldn’t we filter out the records that don’t match the QNAME before counting them?

Author: Changed to require all to have correct owner name. Rare case.

   8.  Extract the name server names from the NS records and any address records
       in the additional section.
   9.  Do [DNS Lookup] of name server names (A and AAAA) not already listed in the

>        additional section of the response.

Reviewer: What if any of those lookups fail? (Maybe other test cases catch that and report on these conditions, but if so, it might be nice to add a note.)

Author: Nothing to be done, but I will make it explicit.

   10. For each IP address add the IP address and *Zone Name* to the
       *Remaining Servers* set unless the IP address is already listed in
       *Handled Servers* together with *Zone Name*.
   11. Create "Intermediate Query Name" by copying *Zone name* as start value.
   12. Run a loop processing *Server Address* (jumps back here from the steps
       below).
       1. Extend "Intermediate Query Name" by adding one more label to the left
          by copying the equivalent label from *Child Zone*. (See "Example 1"
          below.)
       2. Create a [DNS queries][DNS Query] with query name
          *Intermediate Query Name* and [query type] SOA
          ("Intermediate SOA query").
       3. [Send] *Intermediate SOA Query* to *Server Address*. (See "Example 2"
          below.)
       4. Output [B01_SERVER_ZONE_ERROR] with query name *Intermediate Query Name*
          and [query type] SOA and name server IP *Server Address* and go to next
          server in *Remaining Servers* if there is no DNS response.

>        5. If the response has an SOA record with owner name

Reviewer: ```suggestion
       5. If the response has exactly one SOA record with owner name
```

          *Intermediate Query Name* in the answer section, with the AA bit
          set and [RCODE Name] NoError then do:
          1. If *Intermediate Query Name* is equal to *Child Zone* then
             1. Save *Server Address* and *Zone Name* to the *Parent Found* set
                and to the *AA SOA Found* set.
             2. Go to next server in *Remaining Servers*.
          2. Else do:

>              1. Create a [DNS queries][DNS Query] with query name

Reviewer: ```suggestion
             1. Create a [DNS query][DNS Query] with query name
```

                *Intermediate Query Name* and [query type] NS
                ("Intermediate NS query").
             2. [Send] *Intermediate NS Query* to *Server Address*.

>              3. Output [B01_SERVER_ZONE_ERROR] with query name

Reviewer: My remarks and suggestions for step 7 also apply here.

Author: I am not sure what you refer to.

Author: Already fixed.

                *Intermediate NS Query* and [query type] NS and name server IP
                *Server Address* and go to next server in *Remaining Servers* if
                one or more of the following matches:
                   * No DNS response.

>                    * [RCODE Name] besides NoError in response.

Reviewer: ```suggestion
                   * [RCODE Name] different from NoError in response.
```
